lib/core/application/backend.ts

```diff
@@ -179,6 +180,70 @@ class ControllerManager extends ApplicationManager {
       throw runtimeError.get('already_started', 'controller');
     }
 
+    this._add(name, definition);
+  }
+
+  /**
+   * Uses a new controller class.
+   *
+   * The controller class must:
+   *  - call the super constructor with the application instance
+   *  - extend the "Controller" class
+   *  - define the "definition" property
+   *  - (optional) define the "name" property
+   *
+   * The controller name will be inferred from the class name.
+   *   e.g. "PaymentSolutionController" will become "payment-solution"
```

Comment:
I'm not too fond of inferring controller names. Especially considering that the link between the controller class name, and the exposed API entrypoint isn't clear to me.

What about requiring a `name` property in the definition object, and rejecting if it's not there with a meaningful error?

The definition would look like this:

```js
this.definition = {
  name: 'payment',
  actions: {
    // ...
  }
};
```

Reply:
Since it's a class I wanted to have a coherence between the class name and the plugin name. Especially since class name are really important with the new `export` and `import` syntax.

For example our auth plugin class is called `AuthenticationPlugin` (which is a bad generic name) and the plugin name is `kuzzle-plugin-auth-local`. ([see here](https://github.com/kuzzleio/kuzzle-plugin-auth-passport-local/blob/master/lib/index.js#L91))

That's why I wanted to infer the controller name from the class name by default, it will enforce developers to choose meaningful names for their classes.

But for some usecase you may want to set the name by yourself, in this case you use the `name` property:
```
class EmailsController extends Controller {
  constructor (app) {
    super(app);

    this.name = 'emails';

    this.definition = { ... }
  }
}
```

Reply:
The way I see this: when its API is extended with new controllers, Kuzzle needs to know the controller name, its actions, what to do when a controller action is queried, and how those actions can be accessed through HTTP.
With this proposal, 3 of these 4 constraints are regrouped in the `definition` object, and the last one is inferred after the class name. That's why I proposed to also put the name in the definition object.

Your approach might be a valid one if the link between the class name and the controller name is transparent, but I'm not sure this is the case, especially considering that controllers can be added on- the-fly with the `register` method.
Speaking of which, the `register` method takes the controller name as an argument, a closer approach to that behavior with classes might be to pass the controller name to the parent constructor?

One last argument: what about template controllers? Say, an application needs to extend Kuzzle's API, with almost the same controller, with only a few configuration changing. I don't know, might be for multi-tenancy purposes, or to expose the same API for multiple storage indexes with the same structures... if you infer the controller name after the class one, this cannot be done, since Kuzzle will (rightfully) reject new controllers as duplicates.

Reply:
The controller name is inferred from the class name which should most of the use cases.  
About the usecase you are talking about, as I said developers can declare manually the controller name with the `name` property.

Plugins are no longer loaded from the filesystem but they are required/imported from a file or a package. What is imported is a class describing the plugin functionalities so there is no more need for manifest file.

I still didn't had time to work properly on plugins with Kaaf but the manifest file content could be just other properties defined on the plugin class

Reply:
Ah, you're right, it can be done with the `name` property. Ok so I don't see technical reasons to not do it that way. I still don't like it though but, oh, well...